Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp

```diff
@@ -0,0 +1,372 @@
+
+    dbgln_if(ILBM_DEBUG, "decode_iff_chunks");
+
+    while (!chunks.is_empty()) {
+        auto chunk = TRY(decode_iff_advance_chunk(chunks));
+        if (chunk.type == FourCC("CMAP"))
+            context.color_table = TRY(decode_cmap_chunk(chunk));
+        else if (chunk.type == FourCC("BODY"))
+            TRY(decode_body_chunk(chunk, context));
+        else if (chunk.type == FourCC("CRNG"))
+            dbgln_if(ILBM_DEBUG, "Chunk:CRNG");
+    }
+
+    context.state = ILBMLoadingContext::State::ChunksDecoded;
+
+    return {};
+}
+
+static ErrorOr<void> decode_ilbm_first_chunk(ILBMLoadingContext& context)
+{
+    if (context.state >= ILBMLoadingContext::State::BitmapHeaderDecoded)
+        return {};
+
+    context.chunks_cursor = context.data.slice(sizeof(IFF_FILE_HEADER));
+    auto first_chunk = TRY(decode_iff_advance_chunk(context.chunks_cursor));
+
+    if (first_chunk.type != FourCC("BMHD"))
+        return Error::from_string_literal("IFFImageDecoderPlugin: Invalid chunk type, expected BMHD");
+
+    context.bm_header = *bit_cast<BitmapHeader const*>(first_chunk.data.data());
+    context.pitch = ceil_div((u16)context.bm_header.width, (u16)16) * 2;
+    context.state = ILBMLoadingContext::State::BitmapHeaderDecoded;
+
+    dbgln_if(ILBM_DEBUG, "IFFImageDecoderPlugin: BMHD: {}x{} ({},{}), p={}, m={}, c={}",
+        context.bm_header.width,
+        context.bm_header.height,
+        context.bm_header.x,
+        context.bm_header.y,
+        context.bm_header.planes,
+        (u8)context.bm_header.mask,
+        (u8)context.bm_header.compression);
+
+    return {};
+}
+
+ILBMImageDecoderPlugin::ILBMImageDecoderPlugin(ReadonlyBytes data, OwnPtr<ILBMLoadingContext> context)
+    : m_context(move(context))
+{
+    m_context->data = data;
+}
+
+ILBMImageDecoderPlugin::~ILBMImageDecoderPlugin() = default;
+
+IntSize ILBMImageDecoderPlugin::size()
+{
+    return IntSize { m_context->bm_header.width, m_context->bm_header.height };
+}
+
+bool ILBMImageDecoderPlugin::sniff(ReadonlyBytes data)
+{
+
+    ILBMLoadingContext context;
+    context.data = data;
+
+    return !decode_iff_ilbm_header(context).is_error();
+}
+
+ErrorOr<NonnullOwnPtr<ImageDecoderPlugin>> ILBMImageDecoderPlugin::create(ReadonlyBytes data)
+{
+    auto context = TRY(try_make<ILBMLoadingContext>());
+    auto plugin = TRY(adopt_nonnull_own_or_enomem(new (nothrow) ILBMImageDecoderPlugin(data, move(context))));
+    TRY(decode_iff_ilbm_header(*plugin->m_context));
+    TRY(decode_ilbm_first_chunk(*plugin->m_context));
+    return plugin;
+}
+
+ErrorOr<ImageFrameDescriptor> ILBMImageDecoderPlugin::frame(size_t index, Optional<IntSize>)
+{
+    if (index > 0)
+        return Error::from_string_literal("ILBMImageDecoderPlugin: frame index must be 0");
+    else {
+        if (m_context->state < ILBMLoadingContext::ChunksDecoded)
+            TRY(decode_iff_chunks(*m_context));
+
+        RefPtr<Gfx::Bitmap> bitmap = TRY(chunky_to_bitmap(*m_context));
```

Comment:
That will be done every time the frame is requested, you only need to do it once